base/shared/src/main/scala/scalaz/data/ilist.scala

```diff
@@ -16,18 +16,20 @@ trait IListModule {
 
 trait IListInstances {
   implicit final def listEq[A](implicit A: Eq[A]): Eq[IList[A]] =
-    instanceOf[EqClass[IList[A]]] { (a1, a2) =>
-      @tailrec def go(l1: IList[A], l2: IList[A]): Boolean =
-        (IList.uncons(l1), IList.uncons(l2)) match {
-          case (Maybe2.Empty2(), Maybe2.Empty2()) => true
-          case (Maybe2.Just2(x, xs), Maybe2.Just2(y, ys)) =>
-            if (A.equal(x, y)) go(xs, ys)
-            else false
-          case _ => false
-        }
-
-      go(a1, a2)
-    }
+    instanceOf(new EqClass[IList[A]] {
```

Comment:
`IList` does not seem to be an `AnyRef` :/

Reply:
None of the ML-style datatypes are.

Reply:
the only ways around this would be
- implement EqClass in Java to work around missing `eq`.
- define a `Fix` as  `type Fix[F[_] <: AnyRef] <:  AnyRef` 

but not sure we want to go that far into ugly...

Reply:
Add a method to cast to `AnyRef`. The existence of `AnyRef` is a blight on the language.